File: Examples/Reification/College.cpp
// Reification/College.cpp

class College {
  public:
    void course() {}
    void thesis() {}
};

class ClownCollege {
  public:
    void pieThrowing() {}
    void tinyTricycleRiding() {}
    void thesis() {}
};

template <class T> void graduate(T t) {
  t.thesis();
}

int main() {
  graduate(College());
  graduate(ClownCollege());
}

File: Examples/Reification/College.kt
// Reification/College.kt
package reification

class College {
  fun course() {}
  fun thesis() {}
}

class ClownCollege {
  fun pieThrowing() {}
  fun tinyTricycleRiding() {}
  fun thesis() {}
}

fun <T> T.graduate(f: T.() -> Unit) = f()

fun main() {
  val c = College()
  val cc = ClownCollege()
  c.graduate(College::thesis)
  cc.graduate(ClownCollege::thesis)
  c.graduate(College::course)
  cc.graduate(ClownCollege::pieThrowing)
  cc.graduate(ClownCollege::tinyTricycleRiding)
}
